Represent reduce expressions using ForallStmt

vasslitvinov opened the pull request:
The compiler used to convert a reduce expression in Chapel source code to a cascade of `if chpl__tryToken` conditionals, each branch having an explicit loop over the standalone/leader-follower/serial iterator(s) with old-style reduce intents. This happened during parsing.

Now a reduce expression is represented with a PRIM_REDUCE from parsing to resolution. During resolution, it is lowered to a ForallStmt node with a new-style reduce intent. The ForallStmt node is lowered during lowerIterators() alongside other ForallStmt nodes.

De-futurizes test/functions/iterators/bharshbarg/multipleYieldStandalone.chpl
De-futurizes test/reductions/thomasvandoren/test/TestMini.chpl
Resolves #7945.
Resolves #10049.
Resolves #11314.

MUST be followed up by #12174.


### Supporting changes and other notes

A ForallStmt is now allowed to invoke a serial iterator when no parallel iterators are found, under certain conditions.

Iterators invoked from ForallStmts are no longer spared the return-by-ref conversion. Instead, they are subject to that conversion uniformly with other functions. This removes some less-elegant code from callDestructors.cpp. `gRetByRefToken` is added to assist. ForallStmt lowering during lowerIterators() is adjusted to handle this representation, ex. `inlineRetArgFunction()`.

The functions `chpl__staticFastFollowCheck` and `chpl__dynamicFastFollowCheck` are now generated on demand, see `buildFastFollowerChecksIfNeeded()`. They are now generated when the corresponding _iteratorRecord type has been created. That way their formal arguments can be given that type explicitly. Before this change resolution relied on those functions' where-clauses to determine applicability. In some cases resolution of the where-clauses would get into infinite recursion, which now it doesn't.

Various other bug fixes.

There is some handling of _toLeader() functions, which is not extended to _toStandalone(). This is because for a loop expression the compiler generates leader/follower iterators and does not generate the standalone iterator. Ultimately, we should eliminate both _toLeader and _toStandalone(), which will obviate the need to have this handling code and any concerns about supporting _toStandalone().

When lowering a loop expression, the compiler used to generate leader/follower iterators unconditionally, in addition to a serial iterator. The leader iterator would contain a call to _toLeader() applied to the iterable expression. This call does not resolve when the iterable expression does not support parallel iteration. This resolution failure would cause the compiler to switch to the serial iterator silently, due to the tryToken conditionals. Which is not an option now because we no longer use tryToken in this context. Instead, the compiler now checks the availability of the leader iterator before generating the leader iterator for the loop expression, see `haveLeader()`. IteratorGroup #11676 is used for this check.

Analogously, `_toLeader()` functions in the modules now have a where-clause that makes them available only when the underlying iterable has a leader iterator. PRIM_HAS_LEADER is used for that.


### Test changes

Most user-defined reduction op classes needed to be updated by adding new methods, at least identity, accumulateOntoState, clone. These operators worked with the old-style reduce intents, which did not require these methods. The new-style reduce intents do. Also, when a reduce op class has several data fields, I outlined them into an "accumulation state" record in most cases. That way the class itself has only a single "value" field, making it easy to work with the UDRI interface required by the new-style reduce intents. My experience shows that the current UDRI interface leads to boilerplate code, making the new UDRI interface even more desirable. These user-defined reduction op classes will need to be changed again for the new UDRI interface, hopefully making them much tidier.

In `test/studies/kmeans/kmeansonepassreduction-norecord2.chpl`, I chose to use tuples instead of an "accumulation state" record. This is to stay with the spirit of the test. Alas, these tuples included arrays and the resulting executable has memory violations #11984.

The test `studies/kmeans/kmeansonepassreduction.chpl` had a think-o, so I replaced it with (the contents of) `kmeansonepassreduction-mystery.chpl`, which was identical except it did not have the think-o. I, then, removed `kmeansonepassreduction-mystery.chpl`.

The test `reductions/thomasvandoren/mini.chpl` / `TestMini.chpl` felt it was a .future/not working due to the confusion of the current UDRI interface. I made this test work following the pattern of other user-defined-reduction tests. This was an easy un-futurization.

I added or modified .future and .bad files to reduce the impact on nightly testing of the regressions I am leaving for after-merge, see below.


### Regressions left for after-merge

In order to make the new representation available on master, the following regressions are not addressed on this branch. Testing failures have been turned into .futures or updated .bad files, to reduce testing disturbance.

* [ ] #12125 memleak increases

* [ ] #12126 a fast follower is no longer created in `stream-block1d-promote.chpl`

* [ ] #12173 EP size D slight multilocale perf regression

* [ ] #12031 compiler crash due to orphan IR type and functions upon a `.type` expression

* [ ] #11793 enable `forall zip(LF(),S())` where LF provides leader-follower and S() provides only the serial iterator

* [ ] #12069 user-defined "isSorted" reduction produces a wrong answer

* [ ] #12032 prototype partial-reduction code does not compile

* [ ] #12121 compiler crashes upon a .future test where it used to compile successfully

* [ ] #12123 compiler no longer issues an error in .future tests

* [ ] #11984 valgrind failure in kmeansonepassreduction-norecord2

### Other follow-up items

* [ ] Remove old-style forall intent implementation in implementForallIntents.cpp, implementForallIntents2.cpp .

* [ ] Remove gTryToken and chpl__tryToken. To do that, we need to get rid of tryToken in array assignment. For that, we want to allow forall loops to invoke serial iterators when parallel iterators are not available, see #11819.

* [ ] Remove the first argument of PRIM_REDUCE_ASSIGN, which is no longer used.

* [ ] Check if labelSerialIterAsParallel() is still needed. Remove it if not.


### Testing

* [x] linux64 -futures
* [x] gasnet -futures
* [x] --baseline
* [x] --fast
* [x] valgrind
* [x] -memleaks -- some memleak increases #12125
* [x] [perf playground](https://chapel-lang.org/perf/chapcs/reduce-exprs) -- no regressions in the "performance tracking" suite
* [x] multi-locale perf run - EP/D slight slow-down (thanks, Elliot!)

mppf commented on compiler/resolution/functionResolution.cpp:
```diff
@@ -6866,6 +6866,21 @@ static void resolveExternVarSymbols()
+  dtNil->symbol->name = gNil->name;
+
+  // we want gRetByRefToken to be passable to 'ref' formals
+  makeRefType(dtRetByRefToken);
```
It seems to me that it might be more reasonable to tell the compiler to not worry about coercing dtRetByRefToken / getting ref to it / etc.

vasslitvinov replied:
I disagree. That code does not need more special cases, IMO.

mppf commented on compiler/resolution/lowerForalls.cpp:
```diff
@@ -1279,4 +1421,6 @@ void lowerForallStmtsInline()
 
   removeDeadIters();
+
+  INT_ASSERT(gRetByRefToken->firstSymExpr() == NULL);
```
I know what this does but it's worth a comment like "Ensure RetByRefToken is no longer used"

mppf commented on compiler/include/symbol.h:
```diff
@@ -721,6 +726,7 @@ extern StringChainHash uniqueStringHash;
 extern Symbol *gNil;
 extern Symbol *gUnknown;
 extern Symbol *gMethodToken;
+extern Symbol *gRetByRefToken;
```
It's kindof weird, so we definately need a comment about what it's for. (Allow calls to functions that return through a reference argument to exist without creating extra temporaries. This is needed especially for ForallStmts). But I really would like a different name, since this variable is not used normally with ReturnByRef functions the way MethodToken is used for methods. Let's call it `gDummyReturnValue`.

mppf replied:
Or maybe `gDummyValue`.

vasslitvinov replied:
I call it `gDummyRef` because it is a reference.

mppf commented on compiler/include/type.h:
```diff
@@ -396,6 +396,7 @@ TYPE_EXTERN PrimitiveType*    dtValue;
 TYPE_EXTERN PrimitiveType*    dtBorrowed;
 TYPE_EXTERN PrimitiveType*    dtUnmanaged;
 TYPE_EXTERN PrimitiveType*    dtMethodToken;
+TYPE_EXTERN PrimitiveType*    dtRetByRefToken;
```
Can we call it `dtDummyReturnValue` ?

mppf commented on compiler/include/ForallStmt.h:
```diff
@@ -70,20 +78,24 @@ class ForallStmt : public Stmt
+  bool           fFromForLoop; // see comment below
+  bool           fFromReduce;
+  bool           fAllowSerialIterator;
+  bool           fRequireSerialIterator;
```
Hopefully your creation of ForallStmts is factored a little bit into functions so that creating one from a ReduceExpr (or another source) sets e.g. fAllowSerialIterator appropriately. In that case maybe you only need functions returning these, and the fields themselves could be private?

mppf commented on compiler/AST/ForallStmt.cpp:
```diff
@@ -361,8 +373,18 @@ static VarSymbol* createAndAddIndexVar(AList& fIterVars, int idxPosition) {
 }
 
+// Support different users of fsDestructureWhenSingleIdxVar().
+static inline VarSymbol* indexNodeToVarSymbol(BaseAST* index) {
```
I generally avoid the term "AST Node" in favor of the AST type. In this case I'd call the routine `indexExprToVarSymbol`.

mppf commented on compiler/AST/ForallStmt.cpp:
```diff
@@ -516,6 +547,45 @@ ForallStmt* ForallStmt::fromForLoop(ForLoop* forLoop) {
+                                       ShadowVarSymbol* svar,
+                                       bool zippered, bool requireSerial)
+{
+  Symbol* ata = new_IntSymbol(-7, INT_SIZE_64); // atavism, ignored
```
wha-huh? Why -7? Maybe 0 would be more reasonable, even if it's not used. And, if it's not used, how do you know an `int` will do? Could you use `gDummyReturnValue` e.g.?

vasslitvinov replied:
I added, as a follow-up item for this PR, the removal of this arg altogether.

Meanwhile I am using -7 in order to draw attention to this ridiculous-ness.
(Yes I am serious.)

mppf commented on compiler/AST/LoopExpr.cpp:
```diff
@@ -415,6 +411,15 @@ static FnSymbol* buildSerialIteratorFn(const char* iteratorName,
   return sifn;
 }
 
+static Expr* buildLIFNwhere(ArgSymbol* lifnTag, ArgSymbol* lifnIterator,
```
What's a LIFN ? I'd go for a longer name, but you definitely need to add a comment.

mppf commented on compiler/AST/build.cpp:
```diff
@@ -1831,107 +1580,13 @@ buildReduceScanPreface2(FnSymbol* fn, Symbol* eltType, Symbol* globalOp,
-  return new CallExpr(new DefExpr(fn), dataExpr);
+  adjustMinMaxReduceOp(opExpr);
+  return new CallExpr(PRIM_REDUCE, opExpr, dataExpr,
+                      zippered ? gTrue : gFalse);
```
Nice!

mppf commented on compiler/AST/foralls.cpp:
```diff
@@ -605,20 +597,30 @@ static void removeOrigIterCall(SymExpr* origSE)
-      }
+      if (otherSE == def->get(1))
+        def->remove(),
+        ok = true;
```
Curly braces please

mppf commented on compiler/AST/foralls.cpp:
```diff
@@ -810,6 +828,26 @@ static void checkForNonIterator(IteratorGroup* igroup, bool gotSA,
 
+// If iterFn is a serial iterator, make it look parallel, as downstream
+// passes depend on that. I.e. label with FLAG_INLINE_ITERATOR.
+// Create a clone to avoid potential impact on non-parallel uses.
```
Is this creating a parallel version or something? Looks really work-around-y. If it's a workaround, maybe just say it's a workaround in the comment?

mppf commented on compiler/AST/foralls.cpp:
```diff
@@ -1216,3 +1226,78 @@ void resolveForallStmts2() {
+  SymExpr*   opSE = toSymExpr(call->get(1)->remove());           // 1st arg
+  SymExpr* dataSE = toSymExpr(call->get(1)->remove());           // 2nd arg
+  bool   zippered = toSymExpr(call->get(1))->symbol() == gTrue;  // 3rd arg
+  bool  reqSerial = false; // vass todo - require in some cases
```
looks like something to address before merging. Maybe the comment can simply be deleted? (I don't see why a reduce expression would require serial iteration). At the very least update the comment and make an issue.

vasslitvinov replied:
I deferred this to #11819.

mppf commented on compiler/AST/primitive.cpp:
```diff
@@ -584,6 +584,7 @@ initPrimitive() {
   prim_def(PRIM_TRY_EXPR, "try-expr", returnInfoFirst);
   prim_def(PRIM_TRYBANG_EXPR, "try!-expr", returnInfoFirst);
   prim_def(PRIM_YIELD, "yield", returnInfoFirst, true);
+  prim_def(PRIM_REDUCE, "reduce", returnInfoVoid, true); // pre-resolve only
```
Can you add a whole comment describing what PRIM_REDUCE is and what its arguments are?

mppf commented on compiler/AST/primitive.cpp:
```diff
@@ -797,6 +798,7 @@ initPrimitive() {
   // BlockStmt::blockInfo - unlocal local block
   prim_def(PRIM_BLOCK_UNLOCAL, "unlocal block", returnInfoVoid);
 
+  prim_def(PRIM_HAS_LEADER, "has leader", returnInfoBool);
```
Please add a comment indicating what the arguments are

mppf commented on modules/internal/ChapelIteratorSupport.chpl:
```diff
@@ -69,10 +69,30 @@ module ChapelIteratorSupport {
   pragma "no doc"
   proc iteratorIndexType(x) type {
+   // If the result is a runtime type, we better avoid invoking
+   // ic.advance() in iteratorIndex().
```
Yeah but we don't, for say a forall expr yielding arrays. Could we just say that any runtime component of the result is invalid?

mppf commented:
Oh yeah, one more thing. I understand after this PR we can remove gTryToken and chpl__tryToken. Is that the case? Maybe make a backlog issue for doing so?

vasslitvinov commented:
@mppf - thank you, and good suggestions.

>  Is it possible to remove implementForallIntents.cpp or combine it into implementForallIntents2.cpp? 

Very likely. This is a goal, in any case. I will look into it after the merge. I added it as a "followup item" in the PR text.

> I understand after this PR we can remove gTryToken and chpl__tryToken. Is that the case? 

Not yet. It is still used in array assignment. I expect us to get rid of it once we implement #11819. I added this also as a "follow up item".

bradcray commented:
Here are my reactions to the post-merge regressions, sorted by most concerning to least:

* I'm guessing #12126 means that we will lose performance for promoted computations?  That seems very worrisome to me, and not like something that I'd be comfortable releasing without strong support from the team.

* The memory leaks in #11984 seem next most worrisome to me given that we've been trying to drive this number down (so it feels like a step in the wrong direction), and it looks like it'll double our current leaks(?).  Maybe it would help me to understand why these leaks are occurring.

* #12031 seems pretty unfortunate and would be worrisome if it went out into the release.  I suppose I could live with it on master for a short time, but it makes me nervous.

* I feel unsettled not knowing what it is about the futures in issues #12121 and #12123 that causes them to crash now when they didn't before.  Is it clear to you?

* issue #11793 seems reasonable to leave as a separate effort (related to the "when is it OK to fall back to serial iterators vs. not" topic), though for the tests in question I wonder whether followers could be provided/updated to keep them working while adding a new future to capture the issue?  I'm guessing it's linearize() that would need a follower iterator?  Could it just forward to `Xs`' follower iterator?

* I'm not particularly worried about issue #12069 or #12032 for now
